File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/TestOp.java
package org.firstinspires.ftc.teamcode.opmodes;

import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;

import org.firstinspires.ftc.teamcode.drivecontrol.Position;
import org.firstinspires.ftc.teamcode.drivecontrol.Robot;
import org.firstinspires.ftc.teamcode.drivecontrol.Vector2D;

@TeleOp(name="test", group="pushBot")
public class TestOp extends OpMode {
    private static final boolean debug = false;

    private Robot robot;
    private boolean resetIMUOnStart = true;

    private boolean slowMode = false;

    private static final double DEAD_BAND_MAG_SLOW = 0.03;
    private static final double DEAD_BAND_MAG = 0.1;

    @Override
    public void init() {
        robot = new Robot(debug, new Position(0, 0), telemetry, hardwareMap);
    }

    public void init_loop() {
        if (gamepad1.y) {
            resetIMUOnStart = false;
        }
    }

    @Override
    public void start() {
        super.start();
        if (resetIMUOnStart) {
            robot.initIMU();
        }
    }

    @Override
    public void loop() {
        robot.updateDrivePositionTracking();
        slowMode = false;

        Vector2D leftStick1 = new Vector2D(gamepad1.left_stick_x, -gamepad1.left_stick_y);
        Vector2D rightStick1 = new Vector2D(gamepad1.right_stick_x, -gamepad1.right_stick_y);

        if (gamepad1.left_trigger > 0.1) {
            leftStick1 = leftStick1.scale((1-Math.abs(gamepad1.left_trigger))*.75);
            rightStick1 = rightStick1.scale(1-Math.abs(gamepad1.left_trigger));
        }

        robot.updateUsingJoysticks(
                checkDeadBand(leftStick1).scale(Math.sqrt(2)),
                checkDeadBand(rightStick1).scale(Math.sqrt(2))
        );
    }

    private Vector2D checkDeadBand(Vector2D j) {
        return j.getMagnitude() > (slowMode ? DEAD_BAND_MAG_SLOW : DEAD_BAND_MAG)
                ? j : new Vector2D(0, 0);
    }

    @Override
    public void stop() {
        super.stop();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/powerplay/opmodes/TestOp.kt
package org.firstinspires.ftc.teamcode.powerplay.opmodes

import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorSimple
import org.firstinspires.ftc.teamcode.powerplay.Robot
import org.firstinspires.ftc.teamcode.util.math.MathUtil

@TeleOp
class TestOp: OpMode() {
//    private val robot = Robot()

    lateinit var fl: DcMotor
    lateinit var fr: DcMotor
    lateinit var bl: DcMotor
    lateinit var br: DcMotor

    lateinit var slides : DcMotor

    var drive = 0.0
    var strafe = 0.0
    var rotate = 0.0

    private fun driveControl() {
        drive = MathUtil.cubicScaling(0.75, -gamepad1.left_stick_y.toDouble()) * 0.85
        strafe = MathUtil.cubicScaling(0.75, gamepad1.left_stick_x.toDouble()) * 0.85
        rotate = gamepad1.right_stick_x.toDouble() * 0.65
        fl.power = drive + strafe + rotate
        fr.power = drive - strafe - rotate
        bl.power = drive - strafe + rotate
        br.power = drive + strafe - rotate
    }

//    private fun clawControl() {
//        if (gamepad1.a) {
//            robot.clawOpened()
//        }
//        if(gamepad1.b) {
//            robot.clawClosed()
//        }
//    }
//
    private fun slidesControl() {
        if (gamepad1.dpad_up) {
            slides.power = 1.0
        }
        else {
            slides.power = 0.0
        }

        if (gamepad1.dpad_down) {
            slides.power = 0.4
        }
        else {
            slides.power = 0.0
        }
    }
//
//    private fun getTelemetry() {
//    }

    override fun init() {
//        robot.init(hardwareMap)
//        robot.reset()

        fl = hardwareMap.get(DcMotor::class.java, "fl")
        fr = hardwareMap.get(DcMotor::class.java, "fr")
        bl = hardwareMap.get(DcMotor::class.java, "bl")
        br = hardwareMap.get(DcMotor::class.java, "br")

        slides = hardwareMap.get(DcMotor::class.java, "Lift")

        fl.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        fr.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        bl.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        br.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE

        fl.direction = DcMotorSimple.Direction.REVERSE
        bl.direction = DcMotorSimple.Direction.REVERSE
    }

    override fun loop() {
        driveControl()
//        clawControl()
        slidesControl()
//        getTelemetry()
//        robot.update()
    }
}